addons/account/models/account_move.py

```diff
@@ -3072,23 +3086,57 @@ def _get_last_sequence_domain(self, relaxed=False):
+        is_staggered_year = last_month != 12 or last_day != 31
+        if is_staggered_year:
+            if self.date > date(self.date.year, last_month, last_day):
+                year_part = "%s-%s" % (self.date.strftime('%y'), (self.date + relativedelta(years=1)).strftime('%y'))
+            else:
+                year_part = "%s-%s" % ((self.date + relativedelta(years=-1)).strftime('%y'), self.date.strftime('%y'))
         # Arbitrarily use annual sequence for sales documents, but monthly
         # sequence for other documents
         if self.journal_id.type in ['sale', 'bank', 'cash']:
-            starting_sequence = "%s/%04d/00000" % (self.journal_id.code, self.date.year)
+            # We reduce short code to 4 characters (0000) in case of staggered
+            # year to avoid too long sequences (see Indian GST rule 46(b) for
+            # example). Note that it's already the case for monthly sequences.
+            starting_sequence = "%s/%s/%s" % (self.journal_id.code, year_part, '0000' if is_staggered_year else '00000')
         else:
-            starting_sequence = "%s/%04d/%02d/0000" % (self.journal_id.code, self.date.year, self.date.month)
+            starting_sequence = "%s/%s/%02d/0000" % (self.journal_id.code, year_part, self.date.month)
         if self.journal_id.refund_sequence and self.move_type in ('out_refund', 'in_refund'):
             starting_sequence = "R" + starting_sequence
         if self.journal_id.payment_sequence and self.payment_id or self.env.context.get('is_payment'):
             starting_sequence = "P" + starting_sequence
         return starting_sequence
 
     def _get_sequence_date_range(self, reset):
+        if reset not in ('year_range', 'year_range_month'):
+            return super()._get_sequence_date_range(reset)
+
+        fiscalyear_last_day = self.company_id.fiscalyear_last_day
+        fiscalyear_last_month = self.company_id.fiscalyear_last_month
+        date_start, date_end = date_utils.get_fiscal_year(self.date, day=fiscalyear_last_day, month=int(fiscalyear_last_month))
+
         if reset == 'year_range':
-            company = self.company_id
-            return date_utils.get_fiscal_year(self.date, day=company.fiscalyear_last_day, month=int(company.fiscalyear_last_month))
-        return super()._get_sequence_date_range(reset)
+            return (date_start, date_end) + (None, None)
+
+        forced_year_range = (date_start.year, date_end.year)
+        month_range = date_utils.get_month(self.date)
+        fiscalyear_last_month_max_day = calendar.monthrange(self.date.year, int(fiscalyear_last_month))[1]
+        # We need to truncate the month if:
+        # - the fiscal year does not end on the last day of the month
+        # - and the move date is part of that month
+        # The sequence date range will be something like 2020-11-01 to
+        # 2020-11-30. But the sequence should be 2019-2020/11/0001 (or
+        # 2020-2021/11/0001), not 2020-2020/11/0001.
```

Comment:
tricky corner case, did you try it ? I got this ValidationError with a fiscal year ending the 15th of March: 
![image](https://github.com/user-attachments/assets/ee8f8368-29ab-4a75-9484-44a86ae2a03b)

EDIT: found the issue, `fiscalyear_last_month` is a string (in my case: `'3'`) and `self.date.month` is an int (in my case: `3`) so we ignore your subsequent code handling it

Reply:
Indeed, good catch.
I fixed it and tried with the vendor bills, everything seems to be behaving as expected, eg if creating a second bill for second half of april (in case last day is 15 april), it will consider the existing one and increment to 0002, same for first part of april in case there is already one for that first half of the month